Improve status block of UnifiedPushServer CRD

grdryn opened the pull request:
JIRA: https://issues.jboss.org/browse/AEROGEAR-9958


## Progress

- [x] Updated status in CRD
- [x] Added manageError function to set status & fire event
- [x] Add secondary resources to map in CR status
- [x] Set "ready" status appropriately

## Checklist:

- [x] Code has been tested locally by PR requester
- [x] Changes have been successfully verified by another team member

## Verification steps

I've got this all already deployed on an OSD 4 dev cluster, so if you'd like to check it out there, let me know, it might save some time :smiley: 

### Upgrade
- [ ] Deploy operator & UPS from master:
```
git checkout master
make install
# wait for everything to become ready
```
- [ ] Upgrade to configuration from this PR
```
git checkout AEROGEAR-9958
sed -i -e "s|quay.io/aerogear/unifiedpush-operator:.*|quay.io/aerogear/unifiedpush-operator:PR-84|g" deploy/operator.yaml
make install
# ^re-applies all CRDs so they've got the new stuff
```

- [ ] Make sure there are no persistent/blocking errors in the operator logs

### Fresh install
- [ ] Delete the UPS instance:
```
kubectl delete ups --all
```

- [ ] Create a new UPS
```
kubectl apply -f deploy/crds/push_v1alpha1_unifiedpushserver_cr.yaml
```

- [ ] Make sure everything comes up as expected, and there aren't any persistent/blocking errors in the operator logs

grdryn committed:
✨ Update UnifiedPushServerStatus to INTLY standard

JIRA: https://issues.jboss.org/browse/AEROGEAR-9958

INTLY standard here, more or less:
https://issues.jboss.org/browse/INTLY-3364?focusedCommentId=13785555#comment-13785555

grdryn committed:
🔨 Copy manageError method from keycloak operator

JIRA: https://issues.jboss.org/browse/AEROGEAR-9958

grdryn committed:
🚧 Add basic "manageSuccess" method

JIRA: https://issues.jboss.org/browse/AEROGEAR-9958

grdryn committed:
✨ List secondary resources in UPS CR status

JIRA: https://issues.jboss.org/browse/AEROGEAR-9958

grdryn committed:
✨ Set "Ready" field of UPS status

JIRA: https://issues.jboss.org/browse/AEROGEAR-9958

This plagiarises some functions from the keycloak operator, & sets the
ready field of the CR status block based on the status of the 2
Deployments (postgresql & UPS) & the UPS route.

grdryn committed:
🐛 Ignore NoKindMatchError for optional resources

The addition of the monitoring resources here in the reconcile
function didn't consider the case where the kinds don't exist on the
cluster.

grdryn commented on pkg/controller/unifiedpushserver/unifiedpushserver_controller.go:
```diff
@@ -7,13 +7,14 @@ import (
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/record"
 
+	"github.com/aerogear/unifiedpush-operator/pkg/apis/push/v1alpha1"
```
hmm, I'll fix this dupe

davidffrench commented on pkg/controller/unifiedpushserver/unifiedpushserver_controller.go:
```diff
@@ -987,7 +1028,7 @@ func (r *ReconcileUnifiedPushServer) Reconcile(request reconcile.Request) (recon
 	})
-	if err != nil {
+	if _, isNoKindMatchError := err.(*meta.NoKindMatchError); err != nil && !isNoKindMatchError {
 		return reconcile.Result{}, err
```
Is there a reason this and the ones below are not using `r. manageError`?

grdryn replied:
Good spot, thanks -- this is new code that showed up after a rebase & I didn't update it (apart from the fix to ignore if it's a NoKindMatchError). I'll update it :+1:

davidffrench reviewed (commented):
This looks really good @grdryn , great work! Just one very small comment

grdryn committed:
🎨 Organise imports a little better

grdryn commented on deploy/crds/push_v1alpha1_unifiedpushserver_crd.yaml:
```diff
@@ -85,10 +85,31 @@ spec:
           required:
           - phase
+          - message
+          - ready
```
@davidffrench I'm thinking that maybe these shouldn't be "required" for this version of the CRD, wdyt? The case I'm thinking of is where someone applies the CRDs but can't actually upgrade the operator immediately for one reason or another (or have to roll back) -- the existing version of the operator won't be able to reconcile the updated CRD if we make these required.

davidffrench replied:
I would love to know what happens in this scenario for 'required' status fields. The scenario you are outlining makes sense to not make them mandatory.

grdryn replied:
I tested it out, and got the following error in one case:

<details>
{"level":"error","ts":1574341351.4474332,"logger":"kubebuilder.controller","msg":"Reconciler error","controller":"unifiedpushserver-controller","request":"mobile-unifiedpush/unifiedpush","error":"UnifiedPushServer.push.aerogear.org \"unifiedpush\" is invalid: []: Invalid value: map[string]interface {}{\"kind\":\"UnifiedPushServer\", \"metadata\":map[string]interface {}{\"clusterName\":\"\", \"annotations\":map[string]interface {}{\"kubectl.kubernetes.io/last-applied-configuration\":\"{\\\"apiVersion\\\":\\\"push.aerogear.org/v1alpha1\\\",\\\"kind\\\":\\\"UnifiedPushServer\\\",\\\"metadata\\\":{\\\"annotations\\\":{},\\\"generation\\\":1,\\\"name\\\":\\\"unifiedpush\\\",\\\"namespace\\\":\\\"mobile-unifiedpush\\\",\\\"resourceVersion\\\":\\\"\\\",\\\"selfLink\\\":\\\"/apis/push.aerogear.org/v1alpha1/namespaces/mobile-unifiedpush/unifiedpushservers/unifiedpush\\\",\\\"uid\\\":\\\"\\\"},\\\"spec\\\":{\\\"useMessageBroker\\\":false}}\\n\"}, \"resourceVersion\":\"245971\", \"name\":\"unifiedpush\", \"namespace\":\"mobile-unifiedpush\", \"uid\":\"2306dd4a-0c5f-11ea-9c2f-16c554726cf3\", \"generation\":1, \"creationTimestamp\":\"2019-11-21T13:02:13Z\"}, \"status\":map[string]interface {}{\"phase\":\"Provisioning\"}, \"spec\":map[string]interface {}{\"useMessageBroker\":false}, \"apiVersion\":\"push.aerogear.org/v1alpha1\"}: validation failure list:\nstatus.message in body is required\nstatus.ready in body is required","stacktrace":"github.com/aerogear/unifiedpush-operator/vendor/github.com/go-logr/zapr.(*zapLogger).Error\n\t/home/jenkins/workspace/unifiedpush-operator_0.2.0/src/github.com/aerogear/unifiedpush-operator/vendor/github.com/go-logr/zapr/zapr.go:128\ngithub.com/aerogear/unifiedpush-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/home/jenkins/workspace/unifiedpush-operator_0.2.0/src/github.com/aerogear/unifiedpush-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:217\ngithub.com/aerogear/unifiedpush-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1\n\t/home/jenkins/workspace/unifiedpush-operator_0.2.0/src/github.com/aerogear/unifiedpush-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:158\ngithub.com/aerogear/unifiedpush-operator/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1\n\t/home/jenkins/workspace/unifiedpush-operator_0.2.0/src/github.com/aerogear/unifiedpush-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133\ngithub.com/aerogear/unifiedpush-operator/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/home/jenkins/workspace/unifiedpush-operator_0.2.0/src/github.com/aerogear/unifiedpush-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134\ngithub.com/aerogear/unifiedpush-operator/vendor/k8s.io/apimachinery/pkg/util/wait.Until\n\t/home/jenkins/workspace/unifiedpush-operator_0.2.0/src/github.com/aerogear/unifiedpush-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88"}
</details>

^ which basically boils down to `validation failure list:\nstatus.message in body is required\nstatus.ready in body is required` if you care to look closely enough.

I'll update these to make them optional :)

secondsun commented:
@grdryn Does this require crc or can it work on minishift?

grdryn commented:
@secondsun I can't think of any reason why it wouldn't work on minishift.

davidffrench commented:
@secondsun Seconding what @grdryn said, no reason it shouldn't work.

The only reason I have come across for it to not work on k8s 1.11 is a structural schema used in the CRD but that doesn't look to be the case here.

secondsun commented:
Code looks good, code runs good, here's the ups cr after all the updates are applied.

```
[summerspittman@localhost unifiedpush-operator]$ oc describe ups
Name:         example-unifiedpushserver
Namespace:    unifiedpush
Labels:       <none>
Annotations:  kubectl.kubernetes.io/last-applied-configuration={"apiVersion":"push.aerogear.org/v1alpha1","kind":"UnifiedPushServer","metadata":{"annotations":{},"name":"example-unifiedpushserver","namespace":"unif...
API Version:  push.aerogear.org/v1alpha1
Kind:         UnifiedPushServer
Metadata:
  Creation Timestamp:  2019-11-21T13:57:43Z
  Generation:          1
  Resource Version:    5083
  Self Link:           /apis/push.aerogear.org/v1alpha1/namespaces/unifiedpush/unifiedpushservers/example-unifiedpushserver
  UID:                 e3b331a8-0c66-11ea-afb5-52540025c695
Spec:
  Use Message Broker:  false
Status:
  Message:  
  Phase:    Reconciling
  Ready:    true
  Secondary Resources:
    Deployment:
      example-unifiedpushserver-postgresql
      example-unifiedpushserver
    Persistent Volume Claim:
      example-unifiedpushserver-postgresql
    Route:
      example-unifiedpushserver-unifiedpush-proxy
    Secret:
      example-unifiedpushserver-postgresql
    Service:
      example-unifiedpushserver-postgresql
      example-unifiedpushserver-unifiedpush-proxy
      example-unifiedpushserver-unifiedpush
    Service Account:
      example-unifiedpushserver
Events:  <none>

```

grdryn commented:
Thanks a lot for the review & verification @secondsun! As mentioned in another comment, I'm going to make a tiny change to just make the new status fields optional, so it's not breaking backwards compatibility with the older version of the operator (I actually thought that I'd already pushed that before lunch, but clearly not). I don't consider that as needing a re-verification though.

grdryn committed:
🐛 Make new status fields optional for compat

JIRA: https://issues.jboss.org/browse/AEROGEAR-9958

grdryn committed:
🐛 Use Deployment readyReplicas to check readiness

JIRA: https://issues.jboss.org/browse/AEROGEAR-9958

The old version checked a condition, which seemed to get set to true
early on (possibly when it successfully created its replicaset), and
deleting all pods didn't unset it.

This change now considers a Deployment "ready" if it has at least one
replica/pod ready. This should be the point at which these pods are
load-balanced by the Service, so even if not _all_ replicas are ready,
there's enough that the application is "ready" and open for business.

Also, I had previously set a property in the reconcile struct to be
able to access the ready status, but that struct is actually just
initialized once, and then re-used for all invocations of its
Reconcile method -- this doesn't help with how I was using it, I
always expect it to be initially true coming into the function for the
boolean logic to work.

grdryn committed:
🏷 Show ready value even when false

JIRA: https://issues.jboss.org/browse/AEROGEAR-9958

Since we've made this an optional status field, one could only see it
in the yaml returned from the apiserver when it was set to true.

This change fixes that by making it a *bool rather than bool.

grdryn committed:
🔧 Add a code/debug make target for using delve

davidffrench reviewed (approved):
New changes LGTM. Great work @grdryn
